ytypes/leaf.go

```diff
@@ -508,117 +509,203 @@ the resulting Bgp_Neighbor_RouteReflector would have field
+// dereferencing any refs.
+func getUnionKindsNotEnums(schema *yang.Entry) ([]yang.TypeKind, error) {
+	var uks []yang.TypeKind
 	m := make(map[yang.TypeKind]interface{})
-	yts := getUnionTypes(t)
-	for _, yt := range yts {
+	uts, err := getUnionTypesNotEnums(schema, schema.Type)
+	if err != nil {
+		return nil, err
+	}
+	for _, yt := range uts {
 		m[yt.Kind] = nil
 	}
 	for k := range m {
-		out = append(out, k)
+		uks = append(uks, k)
 	}
-	return out
+	return uks, nil
 }
 
-func getUnionTypes(t *yang.YangType) []*yang.YangType {
-	var out []*yang.YangType
-	if t.Kind != yang.Yunion {
-		return []*yang.YangType{t}
-	}
-	for _, t := range t.Type {
-		out = append(out, getUnionTypes(t)...)
+// getUnionTypesNotEnums returns all the non-enum YANG types under the given schema 
+// node, dereferencing any refs.
+func getUnionTypesNotEnums(schema *yang.Entry, yt *yang.YangType) ([]*yang.YangType, error) {
+	var uts []*yang.YangType
+	switch yt.Kind {
+	case yang.Yenum, yang.Yidentityref:
+		// Enum types handled separately.
+		return nil, nil
+	case yang.Yleafref:
+		ns, err := findLeafRefSchema(schema, yt.Path)
+		if err != nil {
+			return nil, err
+		}
+		return getUnionTypesNotEnums(ns, ns.Type)
+	case yang.Yunion:
+		for _, t := range yt.Type {
+			nt, err := getUnionTypesNotEnums(schema, t)
+			if err != nil {
+				return nil, err
+			}
+			uts = append(uts, nt...)
+			if err != nil {
+				return nil, err
+			}
+		}
+	default:
+		uts = []*yang.YangType{yt}
 	}
-	return out
+
+	return uts, nil
 }
 
-func getFieldElemType(parent interface{}, fieldName string) (reflect.Type, error) {
-	t := reflect.TypeOf(parent)
-	ft, ok := t.Elem().FieldByName(fieldName)
-	if !ok {
-		return reflect.TypeOf(nil), fmt.Errorf("%s is not a valid field name in %T", fieldName, parent)
+// schemaToEnumTypes returns the actual enum types (rather than the interface
+// type) for a given schema, which must be for an enum type.
+func schemaToEnumTypes(schema *yang.Entry, t reflect.Type) ([]reflect.Type, error) {
+	enumTypesMethod := reflect.New(t).Elem().MethodByName("ΛEnumTypeMap")
+	if !enumTypesMethod.IsValid() {
+		return nil, fmt.Errorf("type %s does not have a ΛEnumTypesMap function", t)
+	}
+
+	ec := enumTypesMethod.Call(nil)
+	if len(ec) == 0 {
+		return nil, fmt.Errorf("%s ΛEnumTypes function returns empty value", t)
 	}
-	switch {
-	case util.IsTypeStructPtr(t):
-		return ft.Type, nil
-	case util.IsTypeSlicePtr(t):
-		// Dereference slice ptr, then Elem() gives slice element type.
-		return ft.Type.Elem().Elem(), nil
+	ei := ec[0].Interface()
+	enumTypesMap, ok := ei.(map[string][]reflect.Type)
+	if !ok {
+		return nil, fmt.Errorf("%s ΛEnumTypes function returned wrong type %T, want map[string][]reflect.Type", t, ei)
 	}
+	util.DbgPrint("path is %s for schema %s", ygen.EntrySchemaPath(schema), schema.Name)
 
-	return reflect.TypeOf(nil), fmt.Errorf("%T is not a valid parent type", parent)
+	return enumTypesMap[ygen.EntrySchemaPath(schema)], nil
```

Comment:
What happens here if `enumTypeMap` doesn't contain an entry with the relevant path? Should this be a non-error case for the function -- I think not. Consider checking whether the key exists and returning an error if it doesn't.

Reply:
I had originally intended to handle this separately by returning nil but apparently didn't in the end. Good catch.

Reply:
It's not an error. The function that calls this aggregates all possible union values, which can include non-enum types. If this returns nil, it just means that the union doesn't contain enum values.

Reply:
Ah, understood.